pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -1075,6 +1076,45 @@ private void closeConsumerTasks() {
         }
         negativeAcksTracker.close();
         stats.getStatTimeout().ifPresent(Timeout::cancel);
+        releaseMsgIfEnabledPooledMsg();
+    }
+
+    /**
+     * If enabled pooled messages, we should release the messages after closing consumer.
+     *   1. Call "clearIncomingMessages" regardless of whether "internalPinnedExecutor" has shutdown or not.
+     *   2. Increment epoch will auto release all messages which received after close.
+     *   3. Call "clearIncomingMessages" in "internalPinnedExecutor" is used to clear messages in flight.
+     */
+    private void releaseMsgIfEnabledPooledMsg() {
+        if (!poolMessages) {
+            return;
+        }
+        // Increment epoch.
+        CONSUMER_EPOCH.incrementAndGet(this);
+        // Try clear the incoming queue in internalPinnedExecutor-thread.
+        boolean executorIsShutdown = false;
+        try {
+            if (!internalPinnedExecutor.isShutdown()) {
+                internalPinnedExecutor.execute(this::clearIncomingMessages);
+            }
+        } catch (RejectedExecutionException rejectedExecutionException) {
+            executorIsShutdown = true;
+        }
+        // If internalPinnedExecutor is shutdown, try await termination and clear the incoming queue.
+        if (executorIsShutdown) {
+            boolean awaitTerminationSuccess = false;
+            try {
+                awaitTerminationSuccess = internalPinnedExecutor.awaitTermination(1, TimeUnit.SECONDS);
```

Comment:
PulsarClient can create multiple consumers, and `internalExecutorProvider` belongs to `PulsarClient`, so the shutdown of `internalExecutorProvider` should not be done by Consumer

Reply:
the method `awaitTermination` will not shutdown the executor. It works like this:
1. get state, check is terminaled
  1-1. return true if it is terminaled
2. sleep and check the second time.

Reply:
We cannot guarantee that all tasks can be executed after 1 second, so executing `clearIncomingMessages` later cannot guarantee that all messages can be clean.

Reply:
> We cannot guarantee that all tasks can be executed after 1 second

Yes, but there is another mechanism that ensures that:
We increase the consumer epoch before `awaitTermination`, which prevents subsequent messages from entering the queue.

Reply:
These two mechanisms guarantees maximum logical correctness, but is not 100% guaranteed.

Reply:
The message needs to check the epoch before it enters the queue. 
If the consumer is closed, we can set the local epoch to infinite, it can prevent the subsequent messages from entering the queue , then we don’t need to wait for 1 second and check whether executor is shutdown

Reply:
> If the consumer is closed, we can set the local epoch to infinite, it can prevent the subsequent messages from entering the queue , then we don’t need to wait for 1 second and check whether executor is shutdown

If the first task in the task-queue has just finished checking the epoch, but has not yet queued the message, it will not be cleared. Just like this:

| thread: `receive messages` | thread: close consumer |
| --- | --- |
| check the epoch | |
| | set the epoch to infinite |
| | clear the messages queue |
| push messages into the queue |

> we can set the local epoch to infinite

One more that original value dost the same thing as infinite.

Reply:
You are right, because this operation is not overall atomic.
I hope to find a 100% guaranteed. 
Can we disable the `incomingMessages`?
For example: let the size of `GrowableArrayBlockingQueue` become -1, because internal operations are all locked, so atomicity can be maintained.

Reply:
> Can we disable the incomingMessages?
> For example: let the size of GrowableArrayBlockingQueue become -1, because internal operations are all locked, so > atomicity can be maintained.

What a good idea. Already fixed, could you take a look again?